extensions/socketio/src/socketio.server.ts

```diff
@@ -0,0 +1,292 @@
+  ContextView,
+  CoreBindings,
+  CoreTags,
+  createBindingFromClass,
+  inject,
+  MetadataInspector,
+} from '@loopback/core';
+import {HttpServer, HttpServerOptions} from '@loopback/http-server';
+import cors from 'cors';
+import debugFactory from 'debug';
+import {cloneDeep} from 'lodash';
+import SocketIO, {Server, ServerOptions, Socket} from 'socket.io';
+import {
+  getSocketIoMetadata,
+  SocketIoMetadata,
+  SOCKET_IO_CONNECT_METADATA,
+  SOCKET_IO_METADATA,
+  SOCKET_IO_SUBSCRIBE_METADATA,
+} from './decorators';
+import {SocketIoBindings, SocketIoTags} from './keys';
+import {SocketIoControllerFactory} from './socketio-controller-factory';
+const debug = debugFactory('loopback:socketio:server');
+
+export type SockIOMiddleware = (
+  socket: Socket,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  fn: (err?: any) => void,
+) => void;
+
+export const NAMESPACE_KEY_FORMAT = `socketio.namespace.[META_NAME]`;
+
+export const getNamespaceKeyForName = (name: string) =>
+  NAMESPACE_KEY_FORMAT.split('[META_NAME]').join(name);
+
+/**
+ * A binding filter to match socket.io controllers
+ * @param binding - Binding object
+ */
+export const socketIoControllers: BindingFilter = binding => {
+  // It has to be tagged with `controller`
+  if (!binding.tagNames.includes(CoreTags.CONTROLLER)) return false;
+  // It can be explicitly tagged with `socket.io`
+  if (binding.tagNames.includes(SocketIoTags.SOCKET_IO)) return true;
+
+  // Now inspect socket.io decorations
+  if (binding.valueConstructor) {
+    const cls = binding.valueConstructor;
+    const classMeta = MetadataInspector.getClassMetadata(
+      SOCKET_IO_METADATA,
+      cls,
+    );
+    if (classMeta != null) {
+      debug('SocketIo metadata found at class %s', cls.name);
+      return true;
+    }
+    const subscribeMeta = MetadataInspector.getAllMethodMetadata(
+      SOCKET_IO_SUBSCRIBE_METADATA,
+      cls.prototype,
+    );
+    if (subscribeMeta != null) {
+      debug('SocketIo subscribe metadata found at methods of %s', cls.name);
+      return true;
+    }
+
+    const connectMeta = MetadataInspector.getAllMethodMetadata(
+      SOCKET_IO_CONNECT_METADATA,
+      cls.prototype,
+    );
+    if (connectMeta != null) {
+      debug('SocketIo connect metadata found at methods of %s', cls.name);
+      return true;
+    }
+  }
+  return false;
+};
+
+// Server config expected from application
+export interface SocketIoServerOptions {
+  httpServerOptions?: HttpServerResolvedOptions;
+  socketIoOptions?: ServerOptions;
+}
+
+/**
+ * A socketio server
+ */
+export class SocketIoServer extends Context {
+  private controllers: ContextView;
+  private httpServer: HttpServer;
+  private readonly io: Server;
+  public readonly config: HttpServerResolvedOptions;
+
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE) public app: Application,
+    @inject(SocketIoBindings.CONFIG, {optional: true})
+    protected options: SocketIoServerOptions = {},
```

Comment:
Is it possible to use `@config` instead?

Reply:
@raymondfeng the options are got from the root app options in params `socketIoOptions` and `httpServerOptions`. For that is the binding key :
```typescript
  export const CONFIG: BindingKey<SocketIoServerOptions> =
    CoreBindings.APPLICATION_CONFIG;
```
Before it was done this way
```typescript
  export const HTTP_OPTIONS: BindingKey<HttpServerOptions> = CoreBindings.APPLICATION_CONFIG.deepProperty(
    'socketio.http',
  );
  export const SERVER_OPTIONS: BindingKey<ServerOptions> = CoreBindings.APPLICATION_CONFIG.deepProperty(
    'socketio.server',
  );
```
With this, `@config` could be used. Do you think we could improve through this way instead?

Reply:
@raymondfeng curent the options need are `httpServerOptions` and `socketIoOptions`. They are extract from the root applications options through binding key:
```ts
export const CONFIG: BindingKey<SocketIoServerOptions> = CoreBindings.APPLICATION_CONFIG;
```
before they were in separated options.
```typescript
export const HTTP_OPTIONS: BindingKey<HttpServerOptions> = CoreBindings.APPLICATION_CONFIG.deepProperty(
  'socketio.http',
);
export const SERVER_OPTIONS: BindingKey<ServerOptions> = CoreBindings.APPLICATION_CONFIG.deepProperty(
  'socketio.server',
);
```
Returning to this way @config('socketio.http') and @config('socketio.server') could be used. Are you agree?

Reply:
We support injection of config from another binding, such as:

```ts
@config({fromBinding: CoreBindings.APPLICATION_INSTANCE, propertyPath: 'socketio'})
```